cert-ldap-login-module-common/src/main/java/org/esbtools/auth/ldap/LdapRolesProvider.java

```diff
@@ -103,49 +127,78 @@ private void initializeFromConfiguration() throws Exception {
         // A flag that indicates whether to use the SO_KEEPALIVE socket option to attempt to more quickly detect when idle TCP connections have been lost or to prevent them from being unexpectedly closed by intermediate network hardware. By default, the SO_KEEPALIVE socket option will be used.
         options.setUseKeepAlive(ldapConfiguration.isKeepAlive());
 
-
-        if(ldapConfiguration.getUseSSL()) {
+        if (ldapConfiguration.getUseSSL()) {
             TrustStoreTrustManager trustStoreTrustManager = new TrustStoreTrustManager(
-                    ldapConfiguration.getTrustStore(),
-                    ldapConfiguration.getTrustStorePassword().toCharArray(),
-                    "JKS",
-                    true);
+                ldapConfiguration.getTrustStore(),
+                ldapConfiguration.getTrustStorePassword().toCharArray(),
+                "JKS",
+                true);
             SSLSocketFactory socketFactory = new SSLUtil(trustStoreTrustManager).createSSLSocketFactory();
 
-            ldapConnection = new LDAPConnection(
-                    socketFactory,
-                    options,
-                    ldapConfiguration.getServer(),
-                    ldapConfiguration.getPort(),
-                    ldapConfiguration.getBindDn(),
-                    ldapConfiguration.getBindDNPwd()
-            );
+            ldapConnection = new LDAPConnection(socketFactory, options);
         } else {
             LOGGER.warn("Not using SSL to connect to ldap. This is very insecure - do not use in prod environments!");
 
-            ldapConnection = new LDAPConnection(
-                    options,
-                    ldapConfiguration.getServer(),
-                    ldapConfiguration.getPort(),
-                    ldapConfiguration.getBindDn(),
-                    ldapConfiguration.getBindDNPwd()
-            );
+            ldapConnection = new LDAPConnection(options);
+        }
+
+        return ldapConnection;
+    }
+
+    private void connectIfNeeded() throws LDAPException {
+        if (connectionPool != null) {
+            return;
         }
 
-        BindRequest bindRequest = new SimpleBindRequest(ldapConfiguration.getBindDn(), ldapConfiguration.getBindDNPwd());
-        BindResult bindResult = ldapConnection.bind(bindRequest);
+        if (!readyForConnectionAttempt()) {
+            // It's too early to retry connecting again.
+            throw lastSeenConnectionException();
+        }
 
-        if (bindResult.getResultCode() != ResultCode.SUCCESS) {
-            LOGGER.error("Error binding to LDAP" + bindResult.getResultCode());
-            throw new LDAPException(bindResult.getResultCode(), "Error binding to LDAP");
+        if (!attemptingConnect.compareAndSet(/*expect*/ false, /*if false then set to*/ true)) {
```

Comment:
Please consider revising how this is documented, (either with non-inline comments or extracting this out to a method call or something, as I find these inline comments vary hard to read.

Reply:
Okay, so the comment is awesome, and describes very well the iceberg of reasoning under this one line, but it still has the /* inline comments */

Maybe I am alone in being bothered by these

Reply:
Haha, I usually inline comment any parameters that are ambiguous (like boolean flags), but if you think it hurts readability in this case I'm fine with removing them.